Add testing and debugging guide page to docs (#39)
Quansight-Labs/free-threaded-compatibility: docs/debugging.md

# Uncovering concurrency issues, testing and debugging

> Until now, the GIL has allowed developers to elide and get away from
> concurrency issues when writing parallel programs, since the GIL ensured that

Reviewer: ```suggestion
Until now, the GIL has allowed developers to ignore C
safety issues when writing parallel programs, since the GIL ensured that
```

> all thread execution got serialized when any of them tried to simultaneously
> write or read a variable defined in the interpreter. Under the new free-threaded

Reviewer: ```suggestion
all thread execution was serialized simultaneous access
to Python objects and state defined in the interpreter. 
Under the new free-threaded
```

> paradigm, developers now must think about concurrency, distribution and
> parallelism constructs that will allow them to exploit the maximum performance
> of their parallel programs.

Reviewer: ```suggestion
safe access to data from multiple C threads. This is particularly important in C extensions that assumed the GIL serialized access to static state in C programs.
```

Usually, concurrency issues arise when two or more threads try to modify the
same value in memory. In Python, this commonly occurs when a class or function
defines shared state, either via an attribute or a variable that can modified

> from each thread execution scope.

Reviewer: Probably worth adding a note about static state in C or Cython extensions after this sentence. Also note that shared python state is "safe" in the sense that concurrent access cannot generate C thread safety issues. Static C state that assumed the GIL prevented concurrent access does not have that guarantee.

The most common issues related to concurrency in the context of free-threaded
Python are either dirty reads/writes to data, unexpected behavior due to
simultaneous access to native libraries that are not thread-safe, and finally,
major runtime crashes due to memory allocation issues and forbidden

> pointer lookups. While the first case depends on the actual implementation of

Reviewer: It's not just memory allocation and pointer issues, it's generic C thread safety issues due to global state in C extensions. Memory or state corruption and random behavior is generally the most common symptom, but it can also it doesn't have to produce crashes.

the algorithm/routine and may produce unintended results, it would not cause
a fatal crash of the interpreter, as opposed to the last two cases.

In order to discover, handle and debug concurrency issues at large, there are
several strategies, which we will summarize next.

> ## Identify shared state objects.

Reviewer 2: I'm not sure this section is needed, since it is covered a lot more extensively in `porting.md`.

Reviewer: Agreed, link to the porting guide for techniques to statically identify shared state in C code.

    # this increases the probability of concurrent access clashes.
    n_threads = 10
    barrier = threading.Barrier(n_threads)

    # This object will be shared by all the threads.
    cls_instance = MyClass(...)

    def closure():
        # Ensure that all threads reach this point before concurrent execution.
        barrier.wait()
        r = cls_instance.call_unsafe()
        return r

    with ThreadPoolExecutor(max_workers=n_threads) as executor:
        futures = [executor.submit(closure) for _ in range(n_threads)]

    results = [f.result() for f in futures]

    # Do something about the results
    assert check_results(results)
```

Given the non-deterministic nature of parallel execution, such tests may pass
from time to time. In order to reliably ensuring their failure under concurrency,
we recommend using `pytest-repeat`, which enables the `--count` flag in the
`pytest` command:

```bash
# Setting PYTHON_GIL=0 ensures that the GIL is effectively disabled.
PYTHON_GIL=0 pytest -x -v --count=100 test_concurrent.py
```

We advise to set `count` in the order of hundreds and even larger, in order to
ensure at least one concurrent clash event.

> ## Debugging tests that depend on native calls

Reviewer 3: In this section it also makes sense to mention that having a debug version of CPython with sources available will make it much easier to debug some issues. Maybe add a link to `Building from source` in `installing_cpython`? And then add another snippet there showing how to keep sources?

```bash
pyenv install --debug --keep 3.13t-dev
```

If your code has native dependencies, either via C/C++ or Cython, `gdb`
(or `lldb`) can be used as follows:

```bash
# Setting PYTHON_GIL=0 ensures that the GIL is effectively disabled.
PYTHON_GIL=0 gdb --args python my_program.py --args ...

# To test under pytest

> PYTHON_GIL=0 gdb --args python -m pytest -x -v test_here.py::TestClass::test_method[arg]

Reviewer: In zsh you'd need to quote the test name I think.

> PYTHON_GIL=0 gdb --args python -m pytest -x -v test_here.py::TestClass::test_method[arg]

Reviewer: ```suggestion
PYTHON_GIL=0 gdb --args python -m pytest -x -v "test_here.py::TestClass::test_method"
```

Also in zsh you'd need to escape the square brackets, so maybe better to not mention that so you can write a portable snippet.

> ```

Reviewer: Here are the invocations I use for lldb:

```
PYTHON_GIL=0 lldb -- /absolute/path/to/python my_program.py
```

I use pyenv, so it actually looks like:

```bash
PYTHON_GIL=0 lldb -- $(pyenv which python)
```

And then to run pytest:

```
PYTHON_GIL=0 lldb -- $(pyenv which python) ($pyenv which pytest)
```

Again with the absolute path to the pytest script. After much experimentation, this is definitely the most reliable way I've found to get lldb to work correctly with python, especially using pyenv and its shell shim tricks.

Reviewer: Maybe also link to the lldb to gdb command map: https://lldb.llvm.org/use/map.html

It's in the lldb docs but it shows how to do tons of things in gdb too.

When Python is run under `gdb`, several Python integration commands will be
available, such commands start with the `py-` prefix. For instance, the `py-bt`
allows to obtain a Python interpreter backtrace whenever the debugger hits a native
frame, this allows to improve the tracking of execution between Python and native

> frames.

Reviewer: lldb has CPython integration but it doesn't work with CPython 3.12 or newer.

a single Cython declaration.

Enabling `cygdb` requires the compilation of Cython sources with the `--gdb`
flag. After the sources are compiled and linked, it can be used as follows:

```bash
# For example, running the tests of scikit-image.
# build/cp313td/ contains the trace files generated by Cython to be compatible
# with cygdb
PYTHON_GIL=0 cygdb build/cp313td/ -- --args python -m  pytest -x -v skimage/
```

Since `cygdb` requires the Python interpreter version used to compile `gdb`
to match the one to be used during the execution of the script, recompiling `gdb`
will be necessary in order to ensure the most complete debugging experience.
We recommend the `gdb` [compilation instructions](https://www.linuxfromscratch.org/blfs/view/svn/general/gdb.html)
provided by the Linux from scratch project.

`cygdb` defines a set of commands prefixed by `cy` that replace the usual `gdb`
commands. For example `cy run` will start the program with the Cython debugging
extensions enabled, `cy break` will define a breakpoint on a function with the
Cython definition name, `cy next` will step over a Cython line, which is equivalent

> to several lines in the produced C code.

Reviewer: neat, I've never actually gotten cygdb to work, I'll have to play with that if I get a chance.
